mypy/checkexpr.py

```diff
@@ -823,20 +823,36 @@ def infer_function_type_arguments_using_context(
         # valid results.
         erased_ctx = replace_meta_vars(ctx, ErasedType())
         ret_type = callable.ret_type
-        if isinstance(ret_type, TypeVarType):
-            if ret_type.values or (not isinstance(ctx, Instance) or
-                                   not ctx.args):
-                # The return type is a type variable. If it has values, we can't easily restrict
-                # type inference to conform to the valid values. If it's unrestricted, we could
-                # infer a too general type for the type variable if we use context, and this could
-                # result in confusing and spurious type errors elsewhere.
-                #
-                # Give up and just use function arguments for type inference. As an exception,
-                # if the context is a generic instance type, actually use it as context, as
-                # this *seems* to usually be the reasonable thing to do.
-                #
-                # See also github issues #462 and #360.
-                ret_type = NoneTyp()
+        if mypy.checker.is_optional(ret_type) and mypy.checker.is_optional(ctx):
+            # If both the context and the return type are optional, unwrap the optional,
+            # since in 99% cases this is what a user expects. In other words, we replace
+            #     Optional[T] <: Optional[int]
+            # with
+            #     T <: int
+            # while the former would infer T <: Optional[int].
+            ret_type = mypy.checker.remove_optional(ret_type)
+            erased_ctx = mypy.checker.remove_optional(erased_ctx)
+            #
+            # TODO: Instead of this hack and the one below, we need to use outer and
+            # inner contexts at the same time. This is however not easy because of two
+            # reasons:
+            #   * We need to support constraints like [1 <: 2, 2 <: X], i.e. with variables
+            #     on both sides. (This is not too hard.)
+            #   * We need to update all the inference "infrastructure", so that all
+            #     variables in an expression are inferred at the same time.
+            #     (And this is hard, also we need to be careful with lambdas that require
+            #     two passes.)
+        if isinstance(ret_type, TypeVarType) and not (isinstance(ctx, Instance) and ctx.args):
+            # Another special case: the return type is a type variable. If it's unrestricted,
+            # we could infer a too general type for the type variable if we use context,
+            # and this could result in confusing and spurious type errors elsewhere.
+            #
+            # Give up and just use function arguments for type inference. As an exception,
+            # if the context is a generic instance type, actually use it as context, as
+            # this *seems* to usually be the reasonable thing to do.
+            #
+            # See also github issues #462 and #360.
+            ret_type = NoneTyp()
```

Comment:
Would it maybe be more correct to temporarily pretend the return type is either Any or the uninhabited type instead of None?

(I bring this up mostly because I'm not sure how exactly `infer_type_arguments` behaves in no-strict-optional vs strict-optional mode -- it feels like there might maybe be some subtle difference in behavior since None inhabits every type in the former mode and doesn't in the latter. So maybe we should set `ret_type` to something that behaves the same way in both modes? Not sure.)

Reply:
I think it doesn't really matter. Having anything that doesn't depend on function type variables will will infer no constraints. Probably the cleanest option is to just return at this point, I will check if this works.